src/base/path.h

```diff
@@ -0,0 +1,109 @@
+/*
+ * Bittorrent Client using Qt and libtorrent.
+ * Copyright (C) 2022  Vladimir Golovnev <glassez@yandex.ru>
+ * Copyright (C) 2012  Christophe Dumez <chris@qbittorrent.org>
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public License
+ * as published by the Free Software Foundation; either version 2
+ * of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ *
+ * In addition, as a special exception, the copyright holders give permission to
+ * link this program with the OpenSSL project's "OpenSSL" library (or with
+ * modified versions of it that use the same license as the "OpenSSL" library),
+ * and distribute the linked executables. You must obey the GNU General Public
+ * License in all respects for all of the code used other than "OpenSSL".  If you
+ * modify file(s), you may extend this exception to your version of the file(s),
+ * but you are not obligated to do so. If you do not wish to do so, delete this
+ * exception statement from your version.
+ */
+
+#pragma once
+
+#include <QMetaType>
+#include <QString>
+
+#include "pathfwd.h"
+
+class Path final
+{
+public:
+    enum class Format
+    {
+        Uniform,
+        Native
+    };
+
+    Path() = default;
+    Path(const Path &other) = default;
+    Path(Path &&other) = default;
+
+    explicit Path(const QString &pathStr);
```

Comment:
I might be drifting off... so I'll ask. What kind of path does it support?
I can see absolute path is ok, what about relative path?

Do you intend to support the following usage? (note this is related to security in WebUI/WebAPI usage)
```c++
Path("/a/b") / Path("../");
Path("/a/b") + QString("/..");
```

Reply:
>I can see absolute path is ok, what about relative path?

Of course, the relative paths are supported as well (a relative path is not only the one that contains `.` or `..`).
As I have already said, my intention (at least at the first stage) is not to introduce anything new, but only to improve the current use case (i.e. to save the caller from having to take care of separators where it does not concern it, and from the basic cleaning of paths from redundant elements). This is basically a routine but very extensive code transformation, which should preserve the current logic (except in cases where obvious bugs are noticed, which can be easily eliminated without breaking anything else). Something more advanced requires additional research, so it can be added later in separate PRs.


> note this is related to security in WebUI/WebAPI usage

It may be better to handle such cases outside of the Path class, so as not to make normal use inconvenient.

Reply:
>which should preserve the current logic (except in cases where obvious bugs are noticed, which can be easily eliminated without breaking anything else). 

This is the concerning part. If this part (relative paths) isn't well-defined, it might cause problems when part of the Path is taking inputs from user/outside world. And I can't be 100% sure whether this is happening now due to large changes..

Reply:
> If this part (relative paths) isn't well-defined

It is well-defined. Path class support relative path. It just remove redundancies that doesn't affect the meaning of path (ref. `QDir::cleanPath()`).